### PR TITLE
feat: generate hydro profiles during HIFLD grid-building

### PR DESCRIPTION
**Pull request by danielolsen**
[Pull Request doc](https://breakthrough-energy.github.io/docs/user/git_guide.html#d-pull-request)

### Purpose
Generate hydro profiles at the same time as we build the HIFLD grid, so that we can leverage the plant-specific attributes of each hydro plant to generate better profiles (partial fulfillment of #227). We do this based on hydro data by BA, retried from the EIA at https://www.eia.gov/opendata/qb.php?category=3390101 (via their API).

### What the code is doing
- We add a new module `prereise.gather.hydrodata.eia.fetch_historical`, with a user-facing function `get_generation` that retrieves historical generation by BA for a set of regions, over a selected year.
- We add a new function `build_hydro` which uses these per-BA totals to come up with normalized per-BA profiles, and uses these to build a per-plant hydro profile dataframe.
- We call `build_hydro` from within the orchestration function.

### Testing
- `get_generation` has been thoroughly tested and works as expected.
- `build_hydro` works, although the values that are returned are somewhat unexpected (see https://github.com/Breakthrough-Energy/PreREISE/pull/261#issuecomment-1023454400)
- the broader integration with `create_csvs` works, although because of the data that `build_hydro` returns, we need to clip the profiles to the range [0, 1] before running simulations.

### Usage Examples
```python
import pandas as pd
from prereise.gather.griddata.hifld.data_process.profiles import build_hydro, floatify
from prereise.gather.hydrodata.eia.fetch_historical import get_generation

eia_api_key = YOUR_EIA_API_KEY
hydro_plants = pd.read_csv(PATH_TO_FORM_860_3).query("`Energy Source 1` == 'WAT'")
hydro_plants = hydro_plants.merge(pd.read_csv(PATH_TO_FORM_860_2), on="Plant Code")
hydro_plants["Pmax"] = (
    hydro_plants[['Summer Capacity (MW)', 'Winter Capacity (MW)']]
    .max(axis=1)
    .map(floatify)
)
hydro_plants = hydro_plants.loc[~hydro_plants["Balancing Authority Code"].isnull()]
# Build normalized profiles for each plant
df = build_hydro(eia_api_key, hydro_plants, 2021)
# Get the raw generation numbers for each BA
df2 = get_generation(eia_api_key, 2021, regions=hydro_plants["Balancing Authority Code"].unique())
```

### Time estimate
1 hour. This is a draft PR since it depends on #256, and there are still conversations about how we may want to modify the hydro profile generation process, but the code that retrieves and parses the EIA data has been well-tested and should be pretty close to its final state.

### CODE REVIEW

**Comment by danielolsen** [comment]
As part of this, we may also want to condense some of the hydro generators together. Looking at the EIA Form 860 data for 2019, there are 4,014 unique hydro generators identified, but only 1,452 unique plants for these generators, so I think we can equivalently combine the generators within each plant to reduce the size of the data & optimizations.

**Comment by danielolsen** [comment]
There's definitely something funny going on with the data from EIA. At least for 2020, the 'residual' profile (that we generate for the balancing authorities that don't have data listed) has an average capacity factor of 15%, but ranges from -178% to 382% of 'residual' generation capacity. Many of the individual BAs show similar over- and under-generation compared to their listed capacities, with some going as far as 50x deviation. I think this is some combination of pumped storage hydro plus data quality issues.

The larger balancing authorities seem to be _better_ about this:
- Of the balancing authorities with at least 2 GW of hydro capacity, only TVA and DUK ever go negative (at a minimum or -15% and -53%, respectively), and none go above 100%. CISO and PJM aren't counted here, since the number of missing data points was above the default threshold of 24 hours over the course of the year (although maybe we want to change this threshold).
- Below 2 GW, there are a few BAs that look quite strange:
  - AVA goes from -58x to +29x capacity (-79 GW to +40 GW!)
  - BANC goes up to 47x capacity
  - NWMT goes up to 16x capacity
  - AEC goes up to 13x capacity, but the capacity is quite small.

Full 2020 results are below. `Pmax` is in MW, other columns are generation normalized to `Pmax`. `NaN` values are for BAs which have too much missing data, and whose generation should end up in 'residual'
```
                             Pmax        min      mean        max
Balancing Authority Code
BPAT                      21892.6   0.136713  0.377959   0.633091
residual                  21388.3  -1.779992  0.148188   3.816713
CISO                       8970.9        NaN       NaN        NaN
PJM                        8642.8        NaN       NaN        NaN
TVA                        6822.1  -0.151273  0.386645   0.798434
NYIS                       6003.6   0.195549  0.533272   0.852489
WALC                       5789.9   0.004491  0.096799   0.312959
SOCO                       4880.8   0.010244  0.281042   0.793517
MISO                       4822.4   0.073822  0.275811   0.739881
ISNE                       3768.6   0.000000  0.213839   0.617736
DUK                        3306.4  -0.528974  0.088138   0.753085
SWPP                       3190.0   0.075235  0.415999   0.741693
IPCO                       2222.6   0.141726  0.386204   0.768919
SCL                        1978.3   0.045999  0.347589   0.724359
LDWP                       1956.2        NaN       NaN        NaN
CHPD                       1819.2   0.093997  0.546786   0.808047
SPA                        1774.9   0.019156  0.411422   0.873289
SEPA                       1405.0   0.000000  0.208212   0.872598
AVA                        1358.8 -58.623050  0.396319  29.640124
PACW                       1216.9   0.041088  0.223186   0.496343
GCPD                       1114.0   0.245961  1.025063   2.119390
WAUW                       1019.8   0.000000  0.055581   0.092175
WACM                        959.2   0.005213  0.860014   1.595079
DOPD                        840.0        NaN       NaN        NaN
SCEG                        819.0   0.007326  0.121021   0.703297
TPWR                        783.9   0.000000  0.390204   0.923587
BANC                        705.3   0.001418  0.801710  47.686091
NWMT                        683.6   0.387654  0.661559  16.430661
PGE                         681.2   0.045508  0.283031   5.951262
ERCO                        530.0   0.000000  0.122886   0.598113
PSCO                        381.2   0.000000  0.061163   1.408709
PSEI                        339.6   0.141343  1.597063   2.626620
SRP                         271.2        NaN       NaN        NaN
PACE                        253.4  -0.698500  0.601987   1.799526
CPLE                        232.9  -0.004294  0.169525   0.446544
SC                          227.0        NaN       NaN        NaN
YAD                         223.5        NaN       NaN        NaN
TIDC                        220.0   0.000000  0.139492   0.527273
LGEE                        132.3        NaN       NaN        NaN
IID                          65.8   0.000000  0.362473   1.200608
AECI                         61.0        NaN       NaN        NaN
FPL                          43.5        NaN       NaN        NaN
NBSO                         19.9        NaN       NaN        NaN
NEVP                         12.0   0.083333  0.439350   0.833333
AEC                           8.2   0.000000  0.196866  13.780488
```
2021 is generally better for data coverage (we get CISO and PJM, although we lose MISO), but WALC and AVA get worse, and therefore the residual gets worse as well:
```
                             Pmax         min      mean         max
Balancing Authority Code
BPAT                      21848.6    0.138132  0.337359    0.609330
CISO                       8966.9   -0.047397  0.144782    0.600654
PJM                        8628.5    0.019586  0.220204    0.758417
TVA                        6822.1   -0.105833  0.330375    0.807816
residual                   6349.8 -135.122681  0.212012    6.412485
NYIS                       6000.5    0.191484  0.518472    0.832764
WALC                       5789.9    0.001554  0.103265   16.332234
SOCO                       4880.8    0.009630  0.248656    0.694558
MISO                       4822.4         NaN       NaN         NaN
ISNE                       3768.6   -0.000531  0.200736    0.640822
DUK                        3306.4   -0.572224  0.048204    0.767602
SWPP                       3190.0    0.051724  0.349997    0.757680
IPCO                       2222.6    0.147575  0.307242    0.674435
SCL                        1978.3    0.053581  0.348878    4.808674
LDWP                       1956.2   -0.002045  0.047575    0.415602
CHPD                       1819.2    0.062115  0.532367    0.849824
SPA                        1774.9    0.013522  0.366947    0.848499
SEPA                       1405.0    0.000000  0.153310    0.930249
AVA                        1358.8  -28.877686  0.476232  633.678981
PACW                       1216.9   -0.233380  0.215149    0.466760
GCPD                       1114.0    0.253142  0.983355    1.617594
WAUW                       1019.8    0.003922  0.063419    0.089233
WACM                        959.2    0.000000  0.721490    3.878232
DOPD                        840.0    0.059524  0.553524    0.850000
SCEG                        819.0         NaN       NaN         NaN
TPWR                        783.9    0.000000  0.393491    0.918485
BANC                        705.3    0.049624  0.488543    2.309655
NWMT                        683.6    0.330603  0.565429    0.911352
PGE                         677.5    0.036900  0.258846    0.591882
ERCO                        530.0    0.000000  0.104058    0.647170
PSCO                        381.2    0.000000  0.060882    0.548269
PSEI                        339.6    0.032391  0.653230    2.373380
SRP                         271.2   -0.648968  0.088397    0.840708
PACE                        253.4    0.090766  0.446755    1.069455
CPLE                        232.9   -0.004294  0.145570    0.446544
SC                          227.0   -0.008811  0.249008    0.938326
YAD                         223.5         NaN       NaN         NaN
TIDC                        220.0         NaN       NaN         NaN
LGEE                        132.3         NaN       NaN         NaN
IID                          63.1    0.000000  0.416375    0.760697
AECI                         61.0         NaN       NaN         NaN
FPL                          43.5         NaN       NaN         NaN
NBSO                         19.9         NaN       NaN         NaN
NEVP                         12.0    0.083333  0.269463    1.666667
AEC                           8.2         NaN       NaN         NaN
```

**Comment by danielolsen** [comment]
The data issues seem to be in only a couple of hours, so I added an anomaly detection function which screens for any reported power >2x the BA's hydro capacity (as reported by EIA) or < -1 * hydro capacity, deletes these, and linearly imputes. This ends up being <5 hours for each BA (except for one BA which is very often above the 2x threshold, so I've skipped imputation there).

With a combination of imputing these anomalies and then re-normalizing the remaining BAs with peak power > EIA capacity, I'm able to get reasonable hydro shapes.

I had been calculating the 'residual' hydro shape by subtracting the found BAs from the `US48` total, but this still produces a strange shape:
image
In the last commit, I changed from this 'residual' approach to a capacity-weight average approach, and the shape looks a lot better:
image
For 2021, there's less than 1 GW of capacity without EIA data anyway (out of about 100 GW total), so having those shapes be a little off probably won't be too much of a problem.

**Comment by danielolsen** [comment]
The individual profiles look fairly good right now, but any profiles that go negative (which several do, due to pumped storage plants) are infeasible with the current REISE.jl formulation, which has constraints for:
- each plant's generation must be non-negative
- hydro plants follow their profiles exactly

For testing network feasibility right now, I'm just clipping the profiles to a lower bound of zero, but eventually we'll want to fix this, either by exempting all hydro plants from the Pmin constraint, or by distinguishing pumped-storage hydro plants from conventional (since we have this via the `Prime Mover` code in EIA 860) and exempting just pumped-storage hydro from the Pmin constraint.

**Comment by danielolsen** [comment]
This has been rebased and re-tested, so is ready for a full review.

File: prereise/gather/griddata/hifld/data_process/profiles.py [hunk 1]
```diff
@@ -100,3 +104,95 @@ def build_wind(wind_plants, download_directory, year):
...
+    :return: (*pandas.Series*) -- data, imputed as applicable.
+    :raise ValueError: if series isn't a 1-D object
+    """
+    if len(series.shape) != 1:
+        raise ValueError("series must be of one-dimentional shape")
+    suspected_anomalies = (series > rel_max * capacity) | (series < rel_min * capacity)
+    output = series.copy()
+    if suspected_anomalies.sum() > max_count or suspected_anomalies.sum() == 0:
+        return output
+    output.loc[suspected_anomalies] = float("nan")
```
Comment (limingzhou2004) [inline prereise/gather/griddata/hifld/data_process/profiles.py:132]: Should we give a warning message if there are too many anomalies?

Comment (rouille) [reply prereise/gather/griddata/hifld/data_process/profiles.py:132]: That would help the user

**Comment by rouille** [comment]
In the `create_csvs` and `create_profiles` functions, do we want the `nrel_email, `nrel_api_key`, `eia_api_key` arguments to be moved into the solar and hydro dictionaries and check within `create_profiles` that the dictionaries have the keys?

File: prereise/gather/griddata/hifld/data_process/profiles.py [hunk 2]
```diff
@@ -100,3 +104,95 @@ def build_wind(wind_plants, download_directory, year):
...
+
+
+def build_hydro(eia_api_key, hydro_plants, year, **hydro_kwargs):
+    """Use plant-level data to build hydro profiles.
+
+    :param str eia_api_key: API key.
+    :param pandas.DataFrame hydro_plants: data frame of hydro generators.
+    :param int/str year: year of data to fetch and build profiles for.
+    :param dict hydro_kwargs: keyword arguments to pass to
+        :func:`from prereise.gather.hydrodata.eia.fetch_historical.get_generation`.
```
Comment (rouille) [inline prereise/gather/griddata/hifld/data_process/profiles.py]: `from` needs to be removed from the link

File: prereise/gather/hydrodata/eia/fetch_historical.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    a full year of hydro data for the given year.
+
+    :param str api_key: EIA api key from <https://www.eia.gov/opendata/register.php>.
+    :param int/str year: year of data to return. Any regions with too many missing
+        entries in this year will be dropped (see ``acceptable_missing``).
+    :param iterable regions: subset of regions to fetch & return. If None, all will be
+        returned.
+    :param int acceptable_missing: allowable number of missing entries, will be linearly
+        imputed. If more than this number are missing in the given ``year``, the region
+        will be dropped.
```
Comment (rouille) [inline prereise/gather/hydrodata/eia/fetch_historical.py]: I believe we are talking about number of missing hours. Correct?

File: prereise/gather/hydrodata/eia/fetch_historical.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    :param bool verbose: whether or not to report the number of missing for each
+        excluded region.
```
Comment (rouille) [inline prereise/gather/hydrodata/eia/fetch_historical.py]: same comment here

File: prereise/gather/hydrodata/eia/fetch_historical.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+    # Impute
+    imputed = linear(
+        full_data.loc[f"{year}-01-01-00":f"{year}-12-31-23", full_year_data_regions],
+        inplace=False,
+    )
+    # Rename columns for brevity. 'Longname (ABV)' becomes simply 'ABV'
+    imputed.columns = [
+        re.search(r"\(([A-Z]*[0-9]*)\)", c).group(1) for c in imputed.columns
+    ]
+    return imputed
```
Comment (rouille) [inline prereise/gather/hydrodata/eia/fetch_historical.py:130]: Do you think this function can be generalize to any fuel type? We do similar things in `prereise.gather.helpers.py` using EIA form 923.

Comment (danielolsen) [reply prereise/gather/hydrodata/eia/fetch_historical.py:130]: It could be done, assuming the same API patterns from EIA, and using a dictionary for `top_level_category_id` instead of a single fixed value (which is hydro-specific). What would we use the resulting data for, validation?

Comment (rouille) [reply prereise/gather/hydrodata/eia/fetch_historical.py:130]: This is where the `get_monthly_net_generation` from the `helpers` module is used:
```
prereise/gather/tests/test_helpers.py:from prereise.gather.helpers import get_monthly_net_generation
prereise/gather/tests/test_helpers.py:def test_get_monthly_net_generation_argument_type():
prereise/gather/tests/test_helpers.py:            get_monthly_net_generation(a[0], a[1], a[2])
prereise/gather/tests/test_helpers.py:def test_get_monthly_net_generation_argument_value():
prereise/gather/tests/test_helpers.py:            get_monthly_net_generation(a[0], a[1], a[2])
prereise/gather/tests/test_helpers.py:def test_get_monthly_net_generation():
prereise/gather/tests/test_helpers.py:        get_monthly_net_generation(state, table, fuel_type) for fuel_type in fuel_types
prereise/gather/hydrodata/eia/demo/eastern_hydro_v3_demo/eastern_hydro_v3_demo.ipynb:    "                                     get_monthly_net_generation)\n",
prereise/gather/hydrodata/eia/demo/eastern_hydro_v3_demo/eastern_hydro_v3_demo.ipynb:    "        eastern_monthly_hyc_rest[state] = get_monthly_net_generation(state, eia_923_form, 'hydro', hps=False)\n",
prereise/gather/hydrodata/eia/demo/eastern_hydro_v3_demo/eastern_hydro_v3_demo.ipynb:    "        total_state_profile = get_monthly_net_generation(state, eia_923_form, 'hydro', hps=False)\n",
prereise/gather/hydrodata/eia/demo/eastern_hydro_v3_demo/eastern_hydro_v3_demo.ipynb:    "        total_state_profile = get_monthly_net_generation(state, eia_923_form, 'hydro', hps=False)\n",
prereise/gather/hydrodata/eia/demo/western_hydro_v2_demo.ipynb:    "from prereise.gather.helpers import get_monthly_net_generation\n",
prereise/gather/hydrodata/eia/demo/western_hydro_v2_demo.ipynb:    "hydro_monthly_generation_ca = get_monthly_net_generation('CA', plant_generation, 'hydro')"
prereise/gather/hydrodata/eia/demo/western_hydro_v2_demo.ipynb:    "hydro_monthly_generation_wa = get_monthly_net_generation('WA', plant_generation, 'hydro')"
prereise/gather/hydrodata/eia/demo/western_hydro_v2_demo.ipynb:    "    hydro_monthly_generation = get_monthly_net_generation(s, plant_generation, 'hydro')\n",
prereise/gather/hydrodata/eia/demo/western_hydro_v2_demo.ipynb:    "hydro_monthly_generation_wy = get_monthly_net_generation('WY', plant_generation, 'hydro')\n",
"hydro_monthly_generation_wy = get_monthly_net_generation('WY', plant_generation, 'hydro')\n",
```
So it used to generate TAMU hydro profiles. But it relies on a single year Excel file, i.e., ugly.

Comment (rouille) [reply prereise/gather/hydrodata/eia/fetch_historical.py:130]: It does not need to be done here but we should keep in mind that we can use the API to get the generation and avoid using EIA orm 963. Do they also have an API for the capacity of the generators (form 860)?

File: prereise/gather/griddata/hifld/data_process/profiles.py (hunk 1 above)
Comment (danielolsen) [reply prereise/gather/griddata/hifld/data_process/profiles.py:132]: Done. An example of the new warning: `Too many suspected anomalies for PSEI, none will be filtered`.

The reason for this edge-case logic is for when there is a large disparity between the maximum capacity from Form 860 and the magnitude of the profiles. In this case, there will be many times where the generation profile is too high, and we don't want to erase these values: we want to use them later to normalize the profile, so that we get the proper shape, not a clipped shape.

File: prereise/gather/hydrodata/eia/fetch_historical.py (hunk 3 above)
Comment (danielolsen) [reply prereise/gather/hydrodata/eia/fetch_historical.py]: Yes. Do you think there's a way to rephrase this to be clearer?

File: prereise/gather/griddata/hifld/data_process/profiles.py (hunk 2 above)
Comment (danielolsen) [reply prereise/gather/griddata/hifld/data_process/profiles.py]: Done.

File: prereise/gather/griddata/hifld/data_process/profiles.py (hunk 1 above)
Comment (rouille) [reply prereise/gather/griddata/hifld/data_process/profiles.py:132]: Makes sense

File: prereise/gather/hydrodata/eia/fetch_historical.py (hunk 3 above)
Comment (rouille) [reply prereise/gather/hydrodata/eia/fetch_historical.py]: allowable number of missing hours (rows). Values will be linearly imputed or region will be dropped depending on number data fetched that are missing. But you are the native speaker

File: prereise/gather/griddata/hifld/data_process/profiles.py [hunk 6]
```diff
@@ -100,3 +104,98 @@ def build_wind(wind_plants, download_directory, year):
...
+    if suspected_anomalies.sum() > max_count:
+        print(f"Too many suspected anomalies for {series.name}, none will be filtered")
+        return output
+    output.loc[suspected_anomalies] = float("nan")
+    return linear(output.to_frame(), inplace=False).squeeze()
+
+
+def build_hydro(eia_api_key, hydro_plants, year, **hydro_kwargs):
+    """Use plant-level data to build hydro profiles.
+
```
Comment (limingzhou2004) [inline prereise/gather/griddata/hifld/data_process/profiles.py]: The current function has the data downloading function and hourly profile processing codes mixed in one method. I wonder if we can extract the hourly profile codes into a separate function, like `produc_hourly_profile()`?  This makes the codes easier to understand and to improve the profile generation method in the future.

File: prereise/gather/griddata/hifld/data_process/profiles.py [hunk 7]
```diff
@@ -100,3 +104,98 @@ def build_wind(wind_plants, download_directory, year):
...
+            template = normalized_profiles[region]
+        else:
+            print(region, "uses 'default' hydro profile")
+            template = normalized_profiles["default"]
+        # pandas won't broadcast for us, so we need to construct a 2D object
+        profiles.loc[:, plant_ids] = pd.concat(
+            [template] * len(plant_ids), axis=1
+        ).values
+
+    return profiles
```
Comment (limingzhou2004) [inline prereise/gather/griddata/hifld/data_process/profiles.py:197]: The implementation of the method looks fine. The assumption that all generators have the same hourly profile may be questioned by users. The load disaggregation has the the similar issue. I will post the question into our slack channel.

File: prereise/gather/hydrodata/eia/fetch_historical.py (hunk 5 above)
Comment (danielolsen) [reply prereise/gather/hydrodata/eia/fetch_historical.py:130]: We've done some thinking about this in the past: https://github.com/Breakthrough-Energy/PreREISE/issues/246

There are potentially several different sources of Form 860 data.

File: prereise/gather/griddata/hifld/data_process/profiles.py (hunk 6 above)
Comment (danielolsen) [reply prereise/gather/griddata/hifld/data_process/profiles.py]: Done.

File: prereise/gather/hydrodata/eia/fetch_historical.py (hunk 3 above)
Comment (danielolsen) [reply prereise/gather/hydrodata/eia/fetch_historical.py]: Done (here and below)

**Comment by danielolsen** [comment]
> In the `create_csvs` and `create_profiles` functions, do we want the `nrel_email`, `nrel_api_key`, `eia_api_key` arguments to be moved into the solar and hydro dictionaries and check within `create_profiles` that the dictionaries have the keys?

This is done in the latest commit. I'll run tests overnight to make sure that everything worked properly.

File: prereise/gather/hydrodata/eia/fetch_historical.py (hunk 3 above)
Comment (rouille) [reply prereise/gather/hydrodata/eia/fetch_historical.py]: Thanks

File: prereise/gather/hydrodata/eia/fetch_historical.py (hunk 5 above)
Comment (rouille) [reply prereise/gather/hydrodata/eia/fetch_historical.py:130]: I am glad we have this API now. We should try to take full advantage of it in the package but this is out of the scope of this PR.

**Review by rouille — APPROVED** [review]
Thanks. It is a nice improvement.

File: prereise/gather/griddata/hifld/data_process/profiles.py [hunk 8]
```diff
@@ -100,3 +104,119 @@ def build_wind(wind_plants, download_directory, year):
...
+        if region in normalized_profiles:
+            template = normalized_profiles[region]
+        else:
+            print(region, "uses 'default' hydro profile")
+            template = normalized_profiles["default"]
+        # pandas won't broadcast for us, so we need to construct a 2D object
+        profiles.loc[:, plant_ids] = pd.concat(
+            [template] * len(plant_ids), axis=1
+        ).values
+
```
Comment (limingzhou2004) [inline prereise/gather/griddata/hifld/data_process/profiles.py:196]: Looks we are missing the return of the profiles.

Comment (danielolsen) [reply prereise/gather/griddata/hifld/data_process/profiles.py:196]: Fixed.
